projects/concurrency/main_test.go

```diff
@@ -0,0 +1,184 @@
+
+		if !ok {
+			t.Errorf("key %v should now exist in the cache but does not", insertKey)
+		}
+
+		// (!) NEED TO LEARN >>> TYPE ASSERTION/CONVERSION
+		// "It performs a type assertion on cacheListElementPointer.Value, checking if the value stored there is a pointer to a CacheValue[string, int] struct."
+		// "If the type assertion is successful, it retrieves the concrete value (the pointer to the CacheValue struct) and assigns it to the cacheValue variable."
+		cacheValue := cacheListElementPointer.Value.(*CacheValue[string, int])
+
+		if cacheValue.value != insertValue {
+			t.Errorf("cacheValue.value: got %v want %v", cacheValue.value, insertValue)
+		}
+	})
+
+	t.Run("add and update a single key/value to an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		insertKey := "a"
+		insertValue := 100
+
+		ok := cache.Put(insertKey, insertValue)
+
+		if ok != false {
+			t.Errorf("found an existing value on the key in the cache, where there should have been no existing value")
+		}
+
+		updatedValue := 200
+
+		ok = cache.Put(insertKey, updatedValue)
+
+		if ok != true {
+			t.Errorf("Put should return true to denote a successful updating of that key's value")
+		}
+
+		valuePointer, ok := cache.Get(insertKey)
+
+		if ok != true {
+			t.Errorf("could not find key %v when there should be one", insertKey)
+		}
+
+		// (!) deference the pointer to a value
+		if *valuePointer != updatedValue {
+			t.Errorf("got %v | want %v", *valuePointer, updatedValue)
+		}
+	})
+
+	t.Run("add 3 keys to an empty cache", func(t *testing.T) {
+		cache := NewCache[string, int](3)
+
+		insertKeyValues := map[string]int{"a": 1, "b": 2, "c": 3}
+
+		for insertKey, insertValue := range insertKeyValues {
+			cache.Put(insertKey, insertValue)
+		}
+
+		for insertKey, insertValue := range insertKeyValues {
+			cacheListElementPointer, ok := cache.entries[insertKey]
+
+			if !ok {
+				t.Errorf("key %v was not found in the cache and should have been", insertKey)
+			}
+
+			// (!) type assertion/conversion
+			cacheValue := cacheListElementPointer.Value.(*CacheValue[string, int])
+
+			if cacheValue.value != insertValue {
+				t.Errorf("cacheValue.value: got %v want %v", cacheValue.value, insertValue)
+			}
+		}
+
+		expectedCacheListOrder := []int{3, 2, 1}
+
+		cacheListElementPointer := cache.list.Front()
+
+		for _, expectedCacheListValue := range expectedCacheListOrder {
+			// (!) type assertion/conversion
+			cacheValue := cacheListElementPointer.Value.(*CacheValue[string, int])
+
+			if cacheValue.value != expectedCacheListValue {
+				t.Errorf("cache list element value: got %v | want %v", cacheValue.value, expectedCacheListValue)
+			}
+
+			cacheListElementPointer = cacheListElementPointer.Next()
+		}
+	})
+}
+
+// note: need to add another test here for 6 keys to an empty cache, to check the LRU order
+
+// this test is failing, probably because of the way i am concurrently accessing the cache
+func TestCacheConcurrency(t *testing.T) {
```

Comment:
What concurrency-safe property of your cache is this test attempting to assert? As far as I can tell, it's testing that:

Given a cache that can hold 3 entries
If 10 goroutines each try to write different values to keys that only that goroutine writes to, they will a short amount of time later be able to read back the value that was written

I don't think this is a property promised by our cache interface. I can imagine a cache which does offer such a property, but it would probably require a concept of a transaction, i.e. you'd say to the cache "I'm about to start giving you a sequence of operations and I want a consistent view of the cache until I tell you I'm done", some code like below:

```go
txn := cache.BeginTransaction()
txn.Put("key-0", 0)
value, ok := txn.Get("key-0")
require.True(t, ok)
require.Equal(t, value, 0)
txn.End()
```

Give some careful thought to what promises we _do_ expect to hold. I think the property you're testing here holds _if the value is present_ (i.e. concurrent writes to _different_ keys shouldn't change independent keys), but it's also valid for the key being looked up to be missing (because three other goroutines did writes between this goroutine's write and read). How could you modify your test to test this property instead?

Another example of a promise I think we do expect is: If a bunch of goroutines do a bunch of writes to different keys, then afterwards exactly three keys should be present in the cache, with the last values written to those keys. One of the tricky things with testing concurrent behaviour is that we don't know _which_ three keys. So we need to write a less restrictive test - we do know some properties of those keys though - we know they should be unique (because our cache promises it can hold three unique keys), we know which keys _could be_ expected (in your example set-up here we know we shouldn't see "key-15", for instance), we know they should have the last values written to them (so in your example set-up here we know the value for "key-0" should be 10, if it's present).

There are also tests we can write e.g. where multiple goroutines are writing different values to the same key - again, we can't write _precise_ assertions on what values we expect to be in the cache, but we can assert on properties of what we do find in the cache.

(Also, if you find yourself interested, try writing the transaction-style cache described above! It's a really interesting problem actually! One of the interesting trade-offs is: Where do you want to block? If someone tries to overwrite a value that some other transaction needs to be able to retrieve, do you want to block the write until the other transaction completes? Do you want to keep two different values for the key which are visible to different transactions - if so do you want to do so inside the cache itself, or do you want each transaction to have a list of "pending" writes? Do you want to reject the write completely? Do you want to allow reads/writes from outside transactions, or only inside them? _All_ of those answers may be valid, and others too, but each suggests a different design of the internals. What examples of use-cases might each of those behaviours be useful for? Some example code which illustrates the problem is below)

```go

cache := NewCache()

var wg sync.WaitGroup

wg.Add(2)

timeBefore := time.Now()

go func() {
    txn := cache.BeginTxn()
    txn.Put("key-0", 0)
    time.Sleep(5 * time.Second)
    value := txn.Get("key-0")
    fmt.Printf("From goroutine 1 expect value 0, got: %d\n", value)
    txn.End()
    wg.Done()
}()

go func() {
    txn := cache.BeginTxn()
    txn.Put("key-0", 1)
    time.Sleep(5 * time.Second)
    value := txn.Get("key-0")
    fmt.Printf("From goroutine 1 expect value 1, got: %d\n", value)
    txn.End()
    wg.Done()
}

wg.Wait()

fmt.Printf("Took %v to wait for two goroutines, value of key-0 was %v\n", time.Now().Sub(timeBefore), cache.Get("key-0"))
```